pkg/workflow/providers/oam/apply.go

```diff
@@ -146,10 +206,82 @@ func lookUpValues(v *value.Value) (*common.ApplicationComponent, *value.Value, s
 	return comp, patcher, clusterName, overrideNamespace, env, nil
 }
 
+func (p *provider) loadDynamicComponent(comp *common.ApplicationComponent) (*common.ApplicationComponent, error) {
+	if comp.Type != "ref-objects" {
```

Comment:
nit: conventionally all the types should be camel-cased, we dont mix camel case and snake-case in one type

Reply:
golang doesnt support "union" primitives, however in kubernetes the conventional way of defining union types is combing a union discriminator and serveral pointers to structs, e.g.:

```
struct {
  Type SomeType # [Foo|Bar]
  Foo *Foo
  Bar *Bar
}
```

see the original proposal about api-union at: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1027-api-unions

the implementation is gonna cover CRD in the foreseeable future, currently all the built-in types in kubernetes has conformed w/ the union structure

Reply:
snake-case for component(or trait/workflow-step) type is a history design in KubeVela. So using camel case here will cause inconsistency.

Reply:
The component type here is not a literal type. Except the new added `ref-objects`, other component type matches a ComponentDefinition where cue is used inside to describe the details of rendering component into resources. So here it is not a union type design.